Explicit dynamics

cbolisetti opened the pull request:
This is a recreation of PR #12730 and an alternative to the formulation in PR #12628. 

This PR demonstrates a formulation of the explicit dynamics solver that does not interfere with the current implementation of ActuallyExplicitEuler. This can be considered option 2 for an explicit dynamics formulation, option 1 being PR #12628 implemented by @hugary1995. This PR is the barebones version of the formulation intended for discussions geared towards a consensus on the design. Currently, the following are included in this PR:

- A CentralDifference timeintegrator object that derives from ActuallyExplicitEuler and overrides the ComputeTimeDerivatives function to include the second derivative according to the Central Difference formulas.
- Note that, since the time derivatives and strains, etc. are computed before the solve, the time derivatives should be computed as u_dot = u_old - u_older, and similarly the strains. The derivatives in the CentralDifference timeintegrator are evaluated accordingly. This required access to a solution state that is 'older than older' (current - 3) and @sveerara made changes to framework to provide that access. The framework changes in this PR correspond to this. 
- We also changed inertia kernels (`InertialForce`, `NodalTranslationalInertia`, `NodalRotationalInertia`) so that the inertia kernels are timeintegrator-agnostic and will provide the right residual and Jacobian regardless of the timeintegrator. 
- We also implemented changes in the inertia kernels for lumped mass calculations, so that correct residuals are calculated for lumped mass solves. 
- Tests include 1D, 2D, and 3D problems in the TensorMechanics module. The results match with the corresponding implicit, Newmark-Beta solutions. The formulation works fine with Dirichlet and Periodic BCs. Tests also include lumped mass solves. To benchmark the lumped mass solves, we created identical models with manually lumped masses using nodal kernels. The implicit and explicit results match pretty well here too. 
- While this formulation does not change the current implementation of ActuallyExplicitEuler, it does, however, require that other kernels (inertia kernels and strain kernels, for example) be evaluated differently when CentralDifference timeintegrator is used. This is not ideal and there might be a more elegant way to do this. Therefore, suggestions and ideas are appreciated.
- I recently discovered that this formulation is not working with Preset BCs. Since Preset and Dirichlet BCs were combined recently, I had to set preset=false in the `FunctionDirichletBCs` in the test input files to get them working. I am not yet sure why this happens. 

Tagging @sveerara @bwspenc @recuero

cbolisetti committed:
Added CentralDifference timeintegrator. Refs #13964.

cbolisetti committed:
Framework changes for CentralDifference timeintegrator. Refs #13964.

cbolisetti committed:
Tests for CentralDifference timeintegrator. Refs #13964

cbolisetti committed:
Changes to inertia kernels to be timeintegrator-agnostic.

cbolisetti committed:
CentralDifference timeintegrator tests for structural dynamics. Refs #13964.

cbolisetti committed:
Documentation for CentralDifference timeintegrator. Refs #13964.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/457830/) on 0ea5abd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14658/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14658/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 784625b6bb1587037a99caa2d2a8a6e0b8bdfdd0`

cbolisetti committed:
Clang format changes.

cbolisetti committed:
Fixed Dirichlet BCs to set preset=false. Tests now pass.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/457867/) on 17321ea wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14658/site)

This comment will be updated on new commits.

loganharbour commented on framework/include/timeintegrators/ActuallyExplicitEuler.h:
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include "ExplicitTimeIntegrator.h"
+// #include "LumpedPreconditioner.h"
```
```suggestion
```

loganharbour commented on framework/src/systems/SystemBase.C:
```diff
@@ -1106,6 +1135,15 @@ SystemBase::copyOldSolutions()
+    for (unsigned int i = _solution_state_size - 1; i > 1; --i)
+      *solutionState(i) = *solutionState(i - 1);
+
+    *solutionState(1) = *currentSolution();
```
Shouldn't this be...
```suggestion
    *solutionState(1) = *solutionOld();
```

cbolisetti replied:
Current and old solutions are equal at this point since the solve hasn't been done.

cbolisetti replied:
@sveerara what do you think?

sveerara replied:
@cbolisetti and @loganharbour, the solution state variable is set to store values in the following format: 
solutionState(0) - current solution at time t
solutionState(1) - solution at t - delta t
solutionState(2) - solution at t- 2*delta t
solutionState(3) - solution at t- 3*delta t ; and so on

So I think when the solution is being shifted backward, solutionState(1) should be set to the solution value at time t, which is current solution. But in the earlier part of this function (line 1102 (old file) or 1132 (edited version)), solutionOld is being set to the currentSolution. So setting solutionState(1) to currentSolution or solutionOld here should return the same result.

loganharbour commented:
I've just looked through the framework stuff in an attempt to understand what's going on with the preset issue we discussed today. I would suggest with the solution state access to just point to the solution, old solution, and older solution vectors for the states 0-2, instead of making more vectors and copying. Then add more as necessary for states 3+. It would make sense for `solutionState(0 to 2)` to work even without sizes set. Or to set the sizes according to which dots are available if the user doesn't set them.

cbolisetti commented:
> I've just looked through the framework stuff in an attempt to understand what's going on with the preset issue we discussed today. I would suggest with the solution state access to just point to the solution, old solution, and older solution vectors for the states 0-2, instead of making more vectors and copying. Then add more as necessary for states 3+. It would make sense for `solutionState(0 to 2)` to work even without sizes set. Or to set the sizes according to which dots are available if the user doesn't set them.

@loganharbour that makes sense.

cbolisetti committed:
Fixed doco and test bugs.

cbolisetti committed:
Addressing comments.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/459833/) on 4796a41 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14658/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14658/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 784625b6bb1587037a99caa2d2a8a6e0b8bdfdd0`

cbolisetti committed:
Clang format changes.

sveerara commented:
@loganharbour, thanks for reviewing this PR. I agree with your comment on not adding extra vectors for solution states 0 to 2 and instead making it point to the existing vectors for current, old and older solutions. 

@cbolisetti, please let me know if you need any help with addressing the comments or if you would like me to make any changes to the framework part. Thanks!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/459897/) on 56de571 : invalidated by @cbolisetti

loganharbour commented:
I've re-written the solution state methods in order to use the current vectors. Only thing I'm stuck on now is if the dot dot, dot old, etc, methods need to be generalized as well. @cbolisetti I've looked around for you today but didn't see you around.

cbolisetti commented:
> I've re-written the solution state methods in order to use the current vectors. Only thing I'm stuck on now is if the dot dot, dot old, etc, methods need to be generalized as well. @cbolisetti I've looked around for you today but didn't see you around.

@loganharbour thats great, thank you. I was busy yesterday. I'll stop by today.

moosebuild commented:
Job [Test timings](https://civet.inl.gov/job/461800/) on 56de571 wanted to post the following:

View timings [here](https://mooseframework.inl.gov/docs/PRs/14658/timing)

This comment will be updated on new commits.

loganharbour committed:
Generalize solution state for obtaining older solutions

refs #13964

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/461959/) on 34ab69d : invalidated by @loganharbour

Writing doc timed out again

loganharbour commented on modules/tensor_mechanics/src/kernels/InertialForce.C:
```diff
@@ -88,23 +102,87 @@ InertialForce::computeQpResidual()
+    Real u_dot_residual_node, u_dotdot_residual_node;
+    for (unsigned int j = 0; j < node.size(); j++)
+    {
+      u_dot_residual_node =
```
`_variable->dofIndices()` can be used here to grab the current indices.

However, accessing these nodal residual values really should be done through some interface if possible.

You shouldn't need to get the nodes and the dof numbers like you are. I believe `_variable->dofIndices()` should be available for the residual calls.

loganharbour commented on modules/tensor_mechanics/src/nodalkernels/NodalTranslationalInertia.C:
```diff
@@ -43,6 +44,7 @@ NodalTranslationalInertia::validParams()
   params.addParam<FileName>(
       "nodal_mass_file",
       "The file containing the nodal positions and the corresponding nodal masses.");
+  params.addParam<bool>("central_difference", false, "Switch for central difference integration.");
```
Where is this used?

loganharbour reviewed (commented):
Only large concern here is the lack of an interface for accessing the nodal residual values

loganharbour commented:
Took care of merge conflicts. I need a review for the last commit (mine), and we should have a discussion about an interface for the nodal residual values as seen in `InertialForce`.

cbolisetti commented:
> Took care of merge conflicts. I need a review for the last commit (mine), and we should have a discussion about an interface for the nodal residual values as seen in `InertialForce`.

Yes, let's talk when you are back next week. One of the tests is failing now though.

loganharbour commented:
> One of the tests is failing now though.

Looks unrelated - passing now. Sounds good for next week.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/51219/) on 1b37911 : invalidated by @loganharbour

Re-try with fixed race conditions

loganharbour commented on modules/tensor_mechanics/src/kernels/InertialForce.C:
```diff
@@ -88,23 +102,87 @@ InertialForce::computeQpResidual()
 }
 
+void
+InertialForce::computeResidual()
```
@dschwen This is the type of madness you're trying to avoid per our conversation yesterday - correct? 

I promised @cbolisetti  I would clean this up some and forgot about it. Now it seems it's back up for discussion after the discussion yesterday.

loganharbour committed:
Use solution vector tag storage for udotfactor and udotdotfactor

refs #13964

moosebuild commented:
Job [External App tests](https://civet.inl.gov/job/526497/) on 947369b : invalidated by @loganharbour

Unrelated failure

loganharbour committed:
Various cleanings for prep for review

- Remove unnecessary line removals/additions
- Add doxygen comments
- Switch pointers to refs where possible

refs #13964

rwcarlsen reviewed (commented):
This is mostly a review of @loganharbour's contributions to this PR - I didn't really look at any of the stuff going in to the tensor mechanics module.

It feels like there is a lot of tricky indexing and handling of the solution state data structures.  I'm lazy and haven't thought enough about it to offer any specific suggestions, but it feels like we could encapsulate the handling of these solution vectors a bit better so everywhere code needs to do stuff with them we don't have to remember how to index and manipulate them correctly.  Also - at least one of my comments is just a general thought - not specific to this PR.

loganharbour committed:
Simplify solution state init

refs #13964

rwcarlsen commented on framework/src/systems/SystemBase.C:
```diff
@@ -107,10 +107,31 @@ SystemBase::SystemBase(SubProblem & subproblem,
+
+  // Create anything that is past older (state of 3+)
+  for (unsigned int i = 3; i <= _old_solution_states_needed; ++i)
+    _solution_states[i] = &addVector("solution_state_" + std::to_string(i), true, GHOSTED);
```
This needs to be modified now that you do lazy setup some object might request 2 older and then another requests 3.  You'll be (re)creating vectors that we already have.

rwcarlsen commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1761,6 +1761,8 @@ class FEProblemBase : public SubProblem, public Restartable
    */
   void uniformRefine();
 
+  virtual void needOldSolutionState(const unsigned int state) override;
```
Is it not possible to get rid of this?  We should't need all the manual forwarding to displaced problem and aux+nonlin systems with the lazy init.

rwcarlsen commented on framework/include/systems/SystemBase.h:
```diff
@@ -806,7 +816,36 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+  /**
+   * Register that a certain solution state is needed (1 = old, 2 = older, etc)
+   */
+  void needOldSolutionState(const unsigned int state);
```
Can we get away with making this private?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/528940/) on cfb18bd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14658/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14658/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7a584baf0930a4788a9e24043eecdf763240abda`

loganharbour committed:
Use solutionState() in all solution(), solutionOld(), etc

refs #13964

loganharbour reviewed (dismissed):
My approval for the commits that aren't mine (modules, modules doco, etc).

loganharbour committed:
Correct index

refs #13964

moosebuild commented:
Job [Private App tests](https://civet.inl.gov/job/530042/) on e875638 : invalidated by @loganharbour

loganharbour reviewed (approved):
My approval for the commits that aren't mine.

hugary1995 commented:
Should CentralDifference and NewmarkBeta stay in framework or in tensor_mechanics?
I am asking because it seems to me that if you don't modify the residual of the second time derivative kernel you won't be able to use those time integrators anyways.
Another way to look at this is CentralDifference and NewmarkBeta are in framework but all regression tests are in tensor_mechanics...

loganharbour commented:
> Should CentralDifference and NewmarkBeta stay in framework or in tensor_mechanics?

I'm fine with moving it - that's probably the best bet here. This has sat around for far too long, I'm going to merge now and will put up a smaller PR for that plus a few other things.

hugary1995 commented:
Okay that makes sense. Thanks to you all @cbolisetti @sveerara @loganharbour I really appreciate all your work!
